File: 2024/src/day01.rs
use std::fs;

pub fn day01() {
	// Part 1.
	let contents = fs::read_to_string("src/inputs/day01-input.txt")
		.expect("Should have been able to read the file");
	let pairs = contents.lines().map(|line| line.split("   ").collect::<Vec<&str>>());
	let mut column1 = pairs.clone().map(|pair| pair[0].parse::<i32>().unwrap()).collect::<Vec<i32>>();
	column1.sort();
	let mut column2 = pairs.clone().map(|pair| pair[1].parse::<i32>().unwrap()).collect::<Vec<i32>>();
	column2.sort();
	let mut distances = Vec::new();
	let mut i = 0;
	for value in column1 {
		distances.insert(i, (value - column2[i]).abs());
		i += 1;
	}
	let mut total_distance = 0;
	for distance in distances {
		total_distance += distance;
	}
	println!("Part 1: {total_distance}");
}

File: 2024/js/day01.mjs
import { getPuzzleInput } from './getPuzzleInput.mjs';

// Part 1.
const lines = (await getPuzzleInput(1)).split('\n').map(line => line.replace(/\s+/, ' '));
lines.pop();
const column1 = lines.map(line => line.split(' ')[0]).map(Number).sort();
const column2 = lines.map(line => line.split(' ')[1]).map(Number).sort();
const distances = column1.map((value, i) => Math.abs(value - column2[i]));
const totalDistance = distances.reduce((a, b) => a + b, 0);
console.info('Part 1:', totalDistance);

// Part 2.
const similarityScores = new Map();
column1.forEach(value => similarityScores.set(value, 0));
for (const value of column1) {
	const count = column2.reduce((a, b) => a + Number(b === value), 0);
	const score = similarityScores.get(value);
	similarityScores.set(value, score + value * count);
}
const totalScore = Array.from(similarityScores).map(value => value[1]).reduce((a, b) => a + b, 0);
console.info('Part 2:', totalScore);
